### ISSUE TITLE
Consider an improved `sf::Clock` API

### ISSUE THREAD

**Issue by ChrisThrasher** [SFML/SFML#2570]
## Subject of the issue

Between [a forum thread of mine](https://en.sfml-dev.org/forums/index.php?topic=28969.0) and the discussion in #2004, there is motivation to find a better API for `sf::Clock`, especially now that it's a more feature-rich class. Now that `sf::Clock` has as many features as it will ever have, it's the prefect time to discuss how to better expose those features in a new API.

Here are a few complaints worth addressing:
 * The name "clock" does not match how the C++ standard uses the word "clock". An `sf::Clock` will not in fact tell you the time. It actually models a stopwatch.
 * Clocks are running by default which is not obvious.
 * `.reset()` does not put the clock in a default-constructed state.
 * You cannot construct a clock in a stopped state.

**Comment by Bromeon** [comment]
From [my previous comment:](https://github.com/SFML/SFML/pull/2004#pullrequestreview-885533903)

> I also believe we don't have to provide every possible combination of operations "for convenience", and we might want to split functionalty. Maybe `start`/`stop` are still justified, as most of the cases, the logic around will know whether the stopwatch should be running or not, and thus `setRunning(bool)` might not be necessary.
> 
> What if we keep the API simple:
> 
> ```c++
> class Clock {
> public:
>     // Running state
>     void start();
>     void stop(); // formerly pause()
>     bool isRunning() const;
> 
>     // Elapsed time state
>     void reset();
>     Time getElapsedTime() const;
> };
> ```
> 
> One option would be to just keep the two states completely orthogonal, so `reset()` would not affect running/stopped state. But in practice, one is likely to either stop the clock or restart it, so maybe `reset()` can have a fixed meaning.
> 
> I'm not sure if we need the methods who act as both setters and getters simultaneously (current `restart()`) -- while convenient, it introduces some redundancy and makes the API surface broader. So far there was only one state (elapsed time), so it was manageable, but now we additionally introduce running/stopped, so we might focus on making sure the user always knows which state he is manipulating.


---

TLDR: a stopwatch has two states that are orthogonal by themselves:
* is running
* elapsed time

And much of the confusion arises by manipulating both at the same time, without having a clear name for it (e.g. `reset` vs. `restart`).

**Comment by MarioLiebisch** [comment]
> Just brainstorming, one option would also be to use assignment instead of setters to re-initialize the stopwatch.

I'm not sure. To me this doesn't feel very intuitive from an object oriented point of view.

How about renaming it to `Timer` instead and – to include the potentially expected functionality – add a new member `sf::Timer::wait(timespan, relative)`?

While this is somewhat covered by `sf::sleep()`, this would allow one to wrap functionality like "do these things, but make sure you'll need at least x seconds", just like the internal frame rate cap.

**Comment by eyalk1** [comment]
Just some of my opinions:
1. ```start``` indicates that you run from the initial value - 0.
     so:
```cpp
tm.start();
...
tm.stop();
...
//Would expect this to reset the clock!
tm.start();
```
    If we change it to ```run```, then ```is_running``` is a bit more correct. 

3. ```reset``` indicates that the value is set to the initial value.
4. ```restart``` indicates that you are starting again from the initial value. So ```restart``` is just ```stop->reset->start```
5. ```stop``` indicates that you don't intend to run again.

So I propose this API:
```cpp
class Clock {
    Time run();
    Time pause();
    bool isRunning();

    // Pauses, resets, and run again, returns the elapsed time. Maybe change to "restart"?
    Time rerun();

    Time getElapsedTime();
    // Doesn't pause! Just sets the value to 0. Also returnes the elapsed time.
    Time reset();
};
```
Having all the functions return the time can save a few calls at the call site, but maybe pollute the code with [[maybe_unused]]. I like it, but it may not be someone else's cup of tea.

**Comment by eyalk1** [comment]
The clock has two states:
1. Is it running.
2. The Value

all the combinations of changing the states are:
| --- | running | pausing | leave the state as is |
|---|---|---|---|
Time=0| ```restart``` (i.e. start again from 0) | ```stop``` | ```reset``` (i.e set the value to 0) |
Time doesn't change | ```continue``` | ```pause``` | no-op |

now one cool thing is that each of these can return the current time (and getting just the time fits pretty snugly into the no-op slot in the table)

```cpp
struct clock{
    Time restart();
    Time stop();
    Time reset();
    Time continue();
    Time pause();
    Time getEstimatedTime();
};
```
wow! a lot of functions, we can fix it like this:
```cpp
enum class RunState {
    Run, Pause, NoChange
};
enum class TimeState {
    Reset,  NoChange
};
struct clock {
    Time operator()(RunState run = NoChange, TimeState time = NoChange){
        // switch statement
    }
    //  this is so if the user only needs to change the timestate he doesn't have to provide a run state
    Time operator()(TimeState time, RunState run = NoChange){ return *this(run,time);}
};
```
is this an overkill? what do you guys think?

**Comment by eyalk1** [comment]
reading documentation is part of the fun! I wouldn't spend more than a minute on reading the documentation for the clock before using it...

We can make the names more explicit:
```cpp
struct clock{
    Time run(); //instead of continue
    Time pause();

    Time resetTime();

    Time runAndResetTime();
    Time pauseAndResetTime(); // "and" makes this sound like a code smell so bad
  
    Time getEstimatedTime();
};
```
These don't look amazing.
we can ask - "is there a subset of these operations that is good enough?"
ofc we can just take ```run```, ``` pause```, ```resetTime``` and ```getEstimatedTime```. And have the user do multiple calls for the combinations at the call site. we can argue that ```pause``` would be coupled with ```resetTime```  most of the time (i.e. pauseAndResetTime :) ) , and if you need to pause without ```resetTime``` you can pretty easily emulate it at the call site.
so maybe:
```cpp
struct clock {
    Time run();

    // this:
    Time Stop(); // does it reset the value? i guess i gotta read the docs...
    // or this:
    Time pause(); // doesn't reset for sure -  so probably call resetTime too at call site...

    Time resetTime();
    Time getEstimateTime();
};
```
    

As of the second case:
I do think that the second case is understandable, albeit non-conventional: 
```cpp
sf::Clock c;
// C++20 using enum:
using sf::Clock::RunStates;
using sf::Clock::TimeStates;
...
// start the clock
Time t = c(Run);
...
// get time
Time t2 = c();
...
// pausing
Time t3 = c(Pause);

Time t4 = c(Run);
Time t5 = c(Reset);
```
just three words! ```Run```, ```Pause``` and ```Reset```! it's simple.

**Comment by ChrisThrasher** [comment]
The team has chosen to not act on this issue in v3 so I’m going to close this issue.

**Comment by kimci86** [comment]
What about keeping the issue open for SFML 4.0.0 milestone?

**Comment by ChrisThrasher** [comment]
At this point I'd rather someone just submit a PR than continue the discussion in here. I don't think this matter is important enough to warrant keeping an issue open for potential a few more years.
